### PR TITLE
Enso color picker

### LINKED ISSUE #9613: Simplify the color picker
Use a single slider with to set LCH Hue value, optionally have preset colors underneath.

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

New color picker, designed for Enso:

https://github.com/enso-org/enso/assets/1047859/c3eff168-6807-4825-b17b-053e3cd8b04c

- Colors never clash: OKLCH lightness and chroma are fixed.
- Easily match colors: Colors of other nodes in the current method are expanded to slices of the color wheel.

Closes #9613.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.

### CODE REVIEW

File: app/gui2/src/components/CircularMenu.vue [hunk 1]
```diff
@@ -153,6 +166,12 @@ const emit = defineEmits<{
   }
 }
 
+.around-circle {
+  position: absolute;
+  width: calc(114px - 10px);
+  height: calc(114px - 10px);
```
Comment (farmaazon) [inline app/gui2/src/components/CircularMenu.vue]: What is 114 and what is 10 here? Maybe comment or variables would make it more readable.

File: app/gui2/src/components/ColorRing.vue [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+}
+
+// === Events ===
+
+function eventAngle(event: MouseEvent) {
+  if (!svgElement.value) return 0
+  const origin = Rect.FromDomRect(svgElement.value.getBoundingClientRect()).center()
+  const offset = Vec2.FromXY(event).sub(origin)
+  return Math.atan2(offset.y, offset.x) / (2 * Math.PI) + 0.25
+}
```
Comment (farmaazon) [inline app/gui2/src/components/ColorRing.vue:83]: Please leave a comment explaining how the angle is measured (from what axis) and in what units (radians?)

File: app/gui2/src/components/ColorRing.vue [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+// If the browser doesn't support OKLCH, the gradient will be specified by computing the number of points specified here
+// in OKLCH, converting to sRGB, and interpolating in HSL. This number has been found to be enough to look close to the
+// intended colors, without excessive gradient complexity (which may affect performance).
+const NONNATIVE_OKLCH_INTERPOLATION_STEPS = 300
+
+const selectedColor = defineModel<string | undefined>()
+const props = defineProps<{
+  standalone: boolean
+  matchableColors: Set<string>
```
Comment (farmaazon) [inline app/gui2/src/components/ColorRing.vue:37]: What does it mean the color is "matchable"? That it's extended to slice?

**Review by farmaazon — CHANGES_REQUESTED** [review]
image

I have a strange offset when running in chrome. Also notice, that at least one slice (on the left in this screenshot) has its gradient going in a wrong way. I think `rangesForInputs` and `cssGradient` should be tested with unit tests.

**Comment by farmaazon** [comment]
Hmm, there is no offset in electron package. But the left/top sides of the circle do not react for mouse hover.

File: app/gui2/src/components/ColorRing.vue (hunk 3 above)
Comment (kazcw) [reply app/gui2/src/components/ColorRing.vue:37]: Yes, this is the set of colors used for other nodes in the same method, that will be made easy to exactly match.

**Comment by kazcw** [comment]
> I have a strange offset when running in chrome.

I have tested in Chrome and have not had this problem. My Chrome is 124.0.6367.92 on M1 Mac.

> Also notice, that at least one slice (on the left in this screenshot) has its gradient going in a wrong way.

That's an optical illusion--each slice is a uniform range of a color used by other nodes in the current method.

**Comment by farmaazon** [comment]
> > I have a strange offset when running in chrome.
> 
> I have tested in Chrome and have not had this problem. My Chrome is 124.0.6367.92 on M1 Mac.

My chrome is 123.0.6312.122. I had problems with newer version, that's why I stuck with this one. I'll give another check.

> That's an optical illusion--each slice is a uniform range of a color used by other nodes in the current method.

You are obviously right :man_facepalming:

**Comment by farmaazon** [comment]
Nope, on 124.0.6367.118 there's still an offset. I'm using linux on wayland gnome.

**Comment by farmaazon** [comment]
In electron package, the behavior is also a bit strange: it gets its proper shape only after moving the mouse a little:

[Screencast from 2024-05-01 15-56-05.webm](https://github.com/enso-org/enso/assets/3919101/60ac1ab3-8b35-430b-8f08-d7309025ec8e)

**Comment by kazcw** [comment]
> Nope, on 124.0.6367.118 there's still an offset. I'm using linux on wayland gnome.

Ok, I'll see if I can reproduce it on a Ubuntu machine

File: app/gui2/src/components/ColorRing.vue (hunk 2 above)
Comment (kazcw) [reply app/gui2/src/components/ColorRing.vue:83]: Added a comment at the top of the module

**Comment by kazcw** [comment]
I was able to reproduce some visual glitches in Electron. I re-implemented the ring shape using HTML/CSS without SVG, which will avoid the browser bugs apparently related to SVG `foreignObject` interacting with CSS gradients.

**Comment by kazcw** [comment]
TODO:
- [x] Unit-test range and gradient computation

**Review by farmaazon — APPROVED** [review]

File: app/gui2/src/components/ColorRing.vue [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+  }
+  return target
+})
+
+// === CSS ===
+
+const cssGradient = computed(() => {
+  const points = gradientPoints(
+    fixedRanges.value,
+    browserSupportsOklch ? 2 : NONNATIVE_OKLCH_INTERPOLATION_STEPS,
```
Comment (Frizi) [inline app/gui2/src/components/ColorRing.vue]: Not sure if this is an adequate enough feature check. `browserSupportsOklch` checks if `oklch` is supported as a color space in general, but not if it is supported as a gradient interpolation method specifically. Firefox notably supports the color space, but [doesn't support interpolation methods in the gradient](https://developer.mozilla.org/en-US/docs/Web/CSS/gradient/conic-gradient#browser_compatibility). This needs specialized checks, something like this:
```js
CSS.supports(`background-image: conic-gradient(in oklch increasing hue, red, blue)`)
CSS.supports(`background-image: conic-gradient(in hsl, red, blue)`)
```
Then, there needs to be a separate codepath where `colorInterpolationMethod` is effectively empty when it is not supported.

File: app/gui2/src/components/ColorRing.vue [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+ *  - This is the axis used by CSS gradients (adjustment is necessary when working with trigonometric functions, which
+ *    start from the positive x-axis).
+ *  - Turns allow constants to be expressed as simple numbers, and can be easily converted to the units used by external
+ *    APIs (radians for math, degrees for culori).
+ */
+
+// If the browser doesn't support OKLCH, the gradient will be specified by computing the number of points specified here
+// in OKLCH, converting to sRGB, and interpolating in HSL. This number has been found to be enough to look close to the
+// intended colors, without excessive gradient complexity (which may affect performance).
+const NONNATIVE_OKLCH_INTERPOLATION_STEPS = 300
```
Comment (Frizi) [inline app/gui2/src/components/ColorRing.vue]: Is 300 really necessary? That sounds like a lot. Given that the fallback is still a gradient, I'd imagine just a few steps of rgb gradient should be really enough to decently approximate the gradient in proper color space. Also, interpolating in `srgb` (the default) is better than in `hsl` in such fallback, as it actually preserves the perceived hue and brightness for gradients of similar colors better.

**Review by Frizi — CHANGES_REQUESTED** [review]
Please test this on firefox as well. Right now the picker gradient is missing:
image

Note that it works after manually adding a flag:

File: app/gui2/src/components/ColorRing.vue (hunk 5 above)
Comment (kazcw) [reply app/gui2/src/components/ColorRing.vue]: Much fewer steps are needed in sRGB.

File: app/gui2/src/components/ColorRing.vue [hunk 6]
```diff
@@ -39,6 +40,10 @@ const emit = defineEmits<{
   close: []
 }>()
 
+const browserSupportsOklchInterpolation = cssSupported(
+  'background-image: conic-gradient(in oklch increasing hue, red, lue)',
```
Comment (Frizi) [inline app/gui2/src/components/ColorRing.vue]: `lue` typo

**Review by Frizi — APPROVED** [review]

**Pull request merged**